File: api-server-node-typescript/src/api/utilization/utilization.controller.ts
import {
  ApiBearerAuth,
  ApiOperation,
  ApiResponse,
  ApiUseTags,
} from '@nestjs/swagger';
import { Controller, Param, Body, Get, Post, UseGuards, NotFoundException } from '@nestjs/common';
import { AuthGuard } from '@nestjs/passport';
import { UtilizationService } from './utilization.service';
import { Utilization } from './utilization.entity';
import { ResponseUtilizationDto, ResponseDeviceDto, RequestUtilizationDto, UtilizationDto } from './utilization.dto';
import { Device } from './../device/device.entity';

@ApiBearerAuth()
@ApiUseTags('Utilization')
@Controller('utilizations')
export class UtilizationController {
  constructor(private readonly utilizationService: UtilizationService) { }

  @Get()
  @UseGuards(AuthGuard('jwt'))
  @ApiOperation({ title: 'Get utilizations' })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  findAll(): Promise<Utilization[]> {
    return this.utilizationService.findAll();
  }

  @Get('/getHighestDevice')
  @UseGuards(AuthGuard('jwt'))
  @ApiOperation({ title: 'Get Highest Device By Each Year' })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  async findHighestDeviceByEachYear(): Promise<ResponseDeviceDto> {
    return new ResponseDeviceDto(await this.utilizationService.findHighestDeviceByEachYear());
  }

  @Get('/getHighestDevice/:year')
  @UseGuards(AuthGuard('jwt'))
  @ApiOperation({ title: 'Get Highest Device By Year' })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  findHighestDeviceByYear(@Param('year') year: number): Promise<ResponseUtilizationDto> {
    return this.utilizationService.findHighestDeviceByYear(year).then(res => {
      return new ResponseUtilizationDto(new UtilizationDto(res.year, res.rate, res.device.device_name));
    }).catch(err => {
      throw new NotFoundException(`${year} year is not exist`);
    });
  }

  @Get('/getHighestRate/:deviceId')
  @UseGuards(AuthGuard('jwt'))
  @ApiOperation({ title: 'Get Highest Rate By deviceId' })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  findHighestRateByYear(@Param('deviceId') deviceId: number): Promise<ResponseUtilizationDto> {
    return this.utilizationService.findHighestRateByDeviceId(new Device(deviceId, null)).then(res => {
      return new ResponseUtilizationDto(new UtilizationDto(res.year, res.rate, res.device.device_name));
    }).catch(err => {
      throw new NotFoundException(`${deviceId} deviceId is not exist`);
    });
  }

  @Post('/predictRate')
  @UseGuards(AuthGuard('jwt'))
  @ApiOperation({ title: 'Predict Rate By DeviceId' })
  @ApiResponse({ status: 401, description: 'Unauthorized' })
  predictRateByDeviceId(@Body() requestUtilizationDto: RequestUtilizationDto): Promise<ResponseUtilizationDto> {
    return this.utilizationService.predictRateByDeviceId(new Device(requestUtilizationDto.device_id, null)).then(res => {
      return new ResponseUtilizationDto(new UtilizationDto(res.year, res.rate, res.device.device_name));
    }).catch(err => {
      throw new NotFoundException(`${requestUtilizationDto.device_id} deviceId is not exist`);
    });
  }
}

File: backend-java/src/main/java/com/banking/api/utilization/controller/UtilizationController.java
package com.banking.api.utilization.controller;

import com.banking.api.utilization.constant.UtilizationConstant;
import com.banking.api.utilization.dto.DevicesResponse;
import com.banking.api.utilization.dto.UtilizationDTO;
import com.banking.api.utilization.dto.UtilizationResponse;
import com.banking.api.utilization.service.UtilizationService;
import com.banking.exception.ResourceNotFoundException;
import com.banking.handler.ResponseHandler;
import io.swagger.annotations.Api;
import io.swagger.annotations.ApiImplicitParam;
import io.swagger.annotations.ApiImplicitParams;
import io.swagger.annotations.ApiOperation;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.Optional;

@Api(tags = {"Utilization"})
@Slf4j
@RestController
@RequestMapping(value = UtilizationConstant.API_URI_PREFIX)
public class UtilizationController {

    @Autowired
    private UtilizationService utilizationService;

    @ApiOperation(value = "/Get Highest Device", notes = "Get Highest Device", response = UtilizationDTO.class, httpMethod = "GET")
    @RequestMapping("/getHighestDevice")
    public DevicesResponse findHighestDeviceByEachYear() {
        return DevicesResponse.builder().devices(utilizationService.findHighestDeviceByEachYear()).build();
    }

    @ApiOperation(value = "Get Highest Device", notes = "Get Highest Device", response = UtilizationDTO.class, httpMethod = "GET")
    @ApiImplicitParams({
            @ApiImplicitParam(name = "year", value = "year", dataType = "long", required = true),
    })
    @RequestMapping("/getHighestDevice/:year")
    public UtilizationResponse findHighestDeviceByYear(@RequestParam long year) {
        return Optional.ofNullable(utilizationService.findHighestDeviceByYear(year)).map(res -> {
            return UtilizationResponse.builder().result(res).build();
        }).orElseThrow(() -> new ResourceNotFoundException("Rate", "year", year));
    }

    @ApiOperation(value = "Get Highest Device", notes = "Get Highest Device", response = UtilizationDTO.class, httpMethod = "GET")
    @ApiImplicitParams({
            @ApiImplicitParam(name = "deviceId", value = "deviceId", dataType = "long", required = true),
    })
    @RequestMapping("/getHighestRate/:deviceId")
    public UtilizationResponse findHighestRateByDeviceId(@RequestParam long deviceId) {
        return Optional.ofNullable(utilizationService.findHighestRateByDeviceId(deviceId)).map(res -> {
            return UtilizationResponse.builder().result(res).build();
        }).orElseThrow(() -> new ResourceNotFoundException("Device", "deviceId", deviceId));
    }

    @ApiOperation(value = "Get Highest Device", notes = "Get Highest Device", response = UtilizationDTO.class, httpMethod = "GET")
    @RequestMapping("/predictRate")
    public ResponseHandler predictRateByDeviceId() {
        return new ResponseHandler("success", utilizationService.findAll());
    }
}
